### PR TITLE
Rework Pillar `top` matching and mine population

### PR DESCRIPTION
**Pull request by NicolasT**

### CODE REVIEW

**Comment by NicolasT** [comment]
@bert-e reset

**Comment by NicolasT** [comment]
@bert-e reset

**Review by TeddyAndrieux — COMMENTED** [review]
Why did you not removed all the `mine.send` from the states ?

For me instead of `mine.send` we should only call `mine.update` it the state just after CAs certificates generations

**Comment by NicolasT** [comment]
> Why did you not removed all the `mine.send` from the states ?
> 
> For me instead of `mine.send` we should only call `mine.update` it the state just after CAs certificates generations

I tried that at first. It doesn't work: the pillar matching relies on roles, and roles are only available once the API server is up, which is not the case when we need access to the certificates for the first time, when bootstrapping the bootstrap node. So, even when running `mine.update`, the `mine_functions` would not be properly set, so no values are pushed, and later `mine.get`s fail.

So, I kept the `mine.send`s for bootstrap sake only.

**Comment by NicolasT** [comment]
@bert-e reset

**Comment by NicolasT** [comment]
@bert-e reset

**Comment by NicolasT** [comment]
@bert-e approve

File: pillar/top.sls.in [hunk 1]
```diff
@@ -1,5 +1,33 @@
...
+
+metalk8s-{{ version }}:
+  # The 'bootstrap' node requires the value of `metalk8s.iso_root_path`, which
+  # is only set in the `metalk8s.roles.bootstrap` Pillar. However, the
+  # compound-match below which assigns that Pillar to a node based on its role
+  # only works when the roles are available, i.e. late in the deployment (when
+  # the API server and associated `ext_pillar`s are in place).
+  # To work-around this, we assign the 'bootstrap' role Pillar data to all nodes
+  # for now, until either a better solution is found, or we get rid of
+  # `iso_root_path` and instead use what comes out of #924.
```
Comment (TeddyAndrieux) [inline pillar/top.sls.in:21]: We can also in the bootstrap script create a pillar file only for the current node assigning this bootstrap roles.

Have something like that in the top.sls
```
metalk8s-{{ version }}:
  '*':
    - {{ grains.id }}
   [...]
```
And then in our bootstrap script we just do 
```echo "{'metalk8s': {'nodes': {'$bootstrap_id': {'roles': ['ca', 'bootstrap', 'master', 'etcd']}}}"```
But then we have an issue with `ext_pillar_first` as this information in the static file will be take over the one in the apiserver...

So I don't really know.
Not really like the `ext_pillar_first` as the information from the apiserver should be taken over the one in the pillar file.

We should maybe get rid of "classic" pillar and only use ext_pillar to set all the information including the mine_functions... and so on.
And then in the master config `ext_pillar` we set the list in the right order we want them to get merged and then with "pillar" args from `def ext_pillar(minion_id, pillar)` we can get the information like roles and also the one from the command line.

Just tested `pillar` arg from ext_pillar contain all the merged informations of the previous ext_pillar + static pillar + **cmd line pillar** so for me it's something we can do

Comment (NicolasT) [reply pillar/top.sls.in:21]: > We can also in the bootstrap script create a pillar file only for the current node assigning this bootstrap roles.

Could even be simpler: pass it as a pillar value on the `salt-call` invocation in the script, like we do for several other pillar values.

Comment (NicolasT) [reply pillar/top.sls.in:21]: >  Not really like the ext_pillar_first as the information from the apiserver should be taken over the one in the pillar file.

Yes. And no. Not using `ext_pillar_first` simply doesn't work: we require the `ext_pillar` values to be able to dispatch on the roles (which are in the pillar...). Basically, when you don't use `ext_pillar_first`), you get crashes on the master side when it tries to compile the pillar top for a minion if said pillar top SLS contains matches on pillar values. So, no way around.

Comment (NicolasT) [reply pillar/top.sls.in:21]: > We should maybe get rid of "classic" pillar and only use ext_pillar to set all the information including the mine_functions... and so on.

Nah. If we can have static data, let's use static data.

**Comment by NicolasT** [comment]
@bert-e reset

**Review by TeddyAndrieux — DISMISSED** [review]
LGTM

**Comment by NicolasT** [comment]
@bert-e reset

**Review by TeddyAndrieux — APPROVED** [review]

**Pull request merged**